File: app/src/main/java/net/ginteam/carmen/contract/company/CompaniesContract.java
package net.ginteam.carmen.contract.company;

import net.ginteam.carmen.contract.FetchContract;
import net.ginteam.carmen.model.Pagination;
import net.ginteam.carmen.model.company.CompanyModel;

import java.util.List;

/**
 * Created by Eugene on 12/27/16.
 */

public interface CompaniesContract {

    interface View extends FetchContract.View {

        void showCompanies(List<CompanyModel> companies, Pagination paginationDetails);

        void showMoreCompanies(List <CompanyModel> companies);

    }

    interface Presenter extends FetchContract.Presenter <View> {

        void selectCompany(CompanyModel company);

        void fetchRecentlyWatchedCompanies();

        void fetchPopularCompanies();

        void fetchFavoriteCompanies();

        void fetchCompaniesForCategory(int categoryId, String filter, int pageNumber);

    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/contract/CompaniesContract.kt
package net.ginteam.carmen.kotlin.contract

import net.ginteam.carmen.kotlin.model.CompanyModel
import net.ginteam.carmen.kotlin.model.PaginationModel

/**
 * Created by eugene_shcherbinock on 2/13/17.
 */
object BaseCompaniesContract {

    interface View : BaseContract.View {

        fun showCompanies(companies: MutableList <CompanyModel>, pagination: PaginationModel? = null)

    }

    interface Presenter <in V : View> : BaseContract.Presenter <V> {

        fun addCompanyToFavorites(company: CompanyModel)
        fun removeCompanyFromFavorites(company: CompanyModel)

    }

}

object CompaniesContract {

    interface View : BaseCompaniesContract.View

    interface Presenter : BaseCompaniesContract.Presenter <View> {

        fun fetchCompanies(categoryId: Int, filterQuery: String = "",
                           sortField: String = "", sortType: String = "", pageNumber: Int = 1)

    }

}

object PopularCompaniesContract {

    interface View : BaseCompaniesContract.View

    interface Presenter : BaseCompaniesContract.Presenter <View> {

        fun fetchPopularCompanies()
        fun getUserCityName(): String

    }

}

object RecentlyWatchedCompaniesContract {

    interface View : BaseCompaniesContract.View

    interface Presenter : BaseCompaniesContract.Presenter <View> {

        fun fetchUserRecentlyWatched()

    }

}

object FavoriteCompaniesContract {

    interface View : BaseCompaniesContract.View

    interface Presenter : BaseCompaniesContract.Presenter <View> {

        fun fetchUserFavorites()

    }

}